diff --git a/modules/secp256k1/module.cpp b/modules/secp256k1/module.cpp
index b98cb2a..48359a2 100644
--- a/modules/secp256k1/module.cpp
+++ b/modules/secp256k1/module.cpp
@@ -122,7 +122,7 @@ end:
                                 CheckRet(secp256k1_context_randomize(ctx, seed.data())) == 1,
                                 "Call to secp256k1_context_randomize failed");
                     }
-                } catch ( fuzzing::datasource::Datasource::OutOfData ) { }
+                } catch ( fuzzing::datasource::Datasource::OutOfData& ) { }
             }
 
             void clone(void) {
@@ -152,7 +152,7 @@ end:
                     if ( ds.Get<bool>() ) {
                         clone();
                     }
-                } catch ( fuzzing::datasource::Datasource::OutOfData ) { }
+                } catch ( fuzzing::datasource::Datasource::OutOfData& ) { }
 
                 return ctx;
             }
@@ -216,7 +216,7 @@ end:
                         if ( ds.Get<bool>() ) {
                             convert();
                         }
-                    } catch ( fuzzing::datasource::Datasource::OutOfData ) { }
+                    } catch ( fuzzing::datasource::Datasource::OutOfData& ) { }
                 }
                 return sig;
             }
@@ -289,7 +289,7 @@ end:
                             }
                         }
                     }
-                } catch ( fuzzing::datasource::Datasource::OutOfData ) { }
+                } catch ( fuzzing::datasource::Datasource::OutOfData& ) { }
 
                 if ( original_size > 0 ) {
                     util::free(data);
@@ -316,7 +316,7 @@ end:
                         if ( ds.Get<bool>() ) {
                             serialize();
                         }
-                    } catch ( fuzzing::datasource::Datasource::OutOfData ) { }
+                    } catch ( fuzzing::datasource::Datasource::OutOfData& ) { }
                 }
                 return pub;
             }
@@ -460,7 +460,7 @@ end:
                         if ( ds.Get<bool>() ) {
                             serializeCompact();
                         }
-                    } catch ( fuzzing::datasource::Datasource::OutOfData ) { }
+                    } catch ( fuzzing::datasource::Datasource::OutOfData& ) { }
                 }
                 return sig;
             }
